File: frontend/amttp_app/lib/core/auth/cross_app_auth_bridge.dart
/// Cross-App Auth Bridge — Flutter Side
///
/// Solves the fundamental problem: Flutter uses SharedPreferences, Next.js
/// uses localStorage. They are **separate storage domains** even on the
/// same origin.
///
/// Solution: When Flutter authenticates, it writes a signed session token
/// to a **document.cookie** (via dart:html on web). Cookies are shared
/// across all paths on the same origin, so Next.js can read them.
///
/// The token is HMAC-SHA256 signed to prevent client-side tampering.
/// The signing key is compiled into the app via --dart-define.
///
/// Flow:
///   1. User signs in via Flutter → AuthService validates credentials
///   2. CrossAppAuthBridge.writeSession() →
///        a. Builds JSON payload {userId, role, mode, exp, iat}
///        b. HMAC-SHA256 signs it → base64url(payload).base64url(signature)
///        c. Writes to document.cookie as `amttp_xauth=<token>; path=/; SameSite=Strict`
///   3. Next.js reads the cookie via middleware or client-side → verifies HMAC → trusts session
///   4. On logout, Flutter clears the cookie
library;

import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'package:flutter/foundation.dart';

// Signing key — override via --dart-define=AUTH_BRIDGE_KEY=<secret>
// In production this MUST be set to a strong random secret shared with Next.js
const String _bridgeKey = String.fromEnvironment(
  'AUTH_BRIDGE_KEY',
  defaultValue: 'amttp-dev-bridge-key-change-in-production-2026',
);

const String kCookieName = 'amttp_xauth';
const Duration kSessionDuration = Duration(hours: 24);

/// Payload embedded in the bridge token
class BridgeTokenPayload {
  final String userId;
  final String email;
  final String role; // e.g. "R3_INSTITUTION_OPS"
  final String mode; // "FOCUS" or "WAR_ROOM"
  final String displayName;
  final int iat; // issued-at (epoch seconds)
  final int exp; // expiry (epoch seconds)

  const BridgeTokenPayload({
    required this.userId,
    required this.email,
    required this.role,
    required this.mode,
    required this.displayName,
    required this.iat,
    required this.exp,
  });

  bool get isExpired => DateTime.now().millisecondsSinceEpoch ~/ 1000 > exp;

  Map<String, dynamic> toJson() => {
        'sub': userId,
        'email': email,
        'role': role,
        'mode': mode,
        'name': displayName,
        'iat': iat,
        'exp': exp,
      };

  factory BridgeTokenPayload.fromJson(Map<String, dynamic> json) {
    return BridgeTokenPayload(
      userId: json['sub'] as String,
      email: json['email'] as String,
      role: json['role'] as String,
      mode: json['mode'] as String,
      displayName: json['name'] as String,
      iat: json['iat'] as int,
      exp: json['exp'] as int,
    );
  }
}

/// Cross-app auth bridge — writes/reads HMAC-signed cookies
class CrossAppAuthBridge {
  CrossAppAuthBridge._();

  // ───────────────────── Token creation ─────────────────────

  /// Create a signed bridge token for the given user session.
  /// Returns `base64url(payload).base64url(hmac)` string.
  static String createToken({
    required String userId,
    required String email,
    required String role,
    required String mode,
    required String displayName,
  }) {
    final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    final payload = BridgeTokenPayload(
      userId: userId,
      email: email,
      role: role,
      mode: mode,
      displayName: displayName,
      iat: now,
      exp: now + kSessionDuration.inSeconds,
    );

    final payloadJson = jsonEncode(payload.toJson());
    final payloadB64 = base64Url.encode(utf8.encode(payloadJson));
    final sig = _sign(payloadB64);
    return '$payloadB64.$sig';
  }

  /// Verify and decode a bridge token. Returns null if invalid/expired.
  static BridgeTokenPayload? verifyToken(String token) {
    try {
      final parts = token.split('.');
      if (parts.length != 2) return null;

      final payloadB64 = parts[0];
      final sig = parts[1];

      // Verify HMAC
      if (_sign(payloadB64) != sig) {
        debugPrint('[CrossAppAuth] HMAC verification failed');
        return null;
      }

      final payloadJson = utf8.decode(base64Url.decode(payloadB64));
      final payload = BridgeTokenPayload.fromJson(
        jsonDecode(payloadJson) as Map<String, dynamic>,
      );

      if (payload.isExpired) {
        debugPrint('[CrossAppAuth] Token expired');
        return null;
      }

      return payload;
    } catch (e) {
      debugPrint('[CrossAppAuth] Token verification error: $e');
      return null;
    }
  }

  // ───────────────────── Cookie I/O (web only) ─────────────────────

  /// Write session cookie so Next.js can read it.
  /// Only works on web platform (uses dart:html via conditional import).
  static void writeCookie(String token) {
    if (!kIsWeb) return;
    _setCookie(
        '$kCookieName=$token; path=/; SameSite=Strict; max-age=${kSessionDuration.inSeconds}');
  }

  /// Clear the session cookie on logout.
  static void clearCookie() {
    if (!kIsWeb) return;
    _setCookie('$kCookieName=; path=/; SameSite=Strict; max-age=0');
  }

  /// Read the bridge token from cookie (used by Flutter to restore session).
  static String? readCookie() {
    if (!kIsWeb) return null;
    return _getCookie(kCookieName);
  }

  // ───────────────────── HMAC helpers ─────────────────────

  static String _sign(String data) {
    final key = utf8.encode(_bridgeKey);
    final bytes = utf8.encode(data);
    final hmacSha256 = Hmac(sha256, key);
    final digest = hmacSha256.convert(bytes);
    return base64Url.encode(digest.bytes);
  }

  // ───────────────────── Platform cookie helpers ─────────────────────
  // These use conditional imports resolved at build time.
  // On non-web platforms they are no-ops.

  static void _setCookie(String value) {
    try {
      // ignore: avoid_web_libraries_in_flutter
      // This import is guarded by kIsWeb check above
      _webSetCookie(value);
    } catch (_) {
      // Not on web — ignore
    }
  }

  static String? _getCookie(String name) {
    try {
      return _webGetCookie(name);
    } catch (_) {
      return null;
    }
  }
}

// ─────────────── Web-specific cookie functions ───────────────
// These are in a separate section to isolate dart:html dependency.
// In production, use conditional imports for multi-platform support.

void _webSetCookie(String value) {
  // Implementation injected at build time via web_cookie_impl.dart
  // Uses document.cookie = value
  _WebCookieHelper.setCookie(value);
}

String? _webGetCookie(String name) {
  return _WebCookieHelper.getCookie(name);
}

/// Web cookie helper — uses document.cookie API
/// This class is only instantiated on web platform
class _WebCookieHelper {
  static void setCookie(String value) {
    if (!kIsWeb) return;
    // Use universal_html or dart:html based on build target
    try {
      // Dynamic invocation to avoid import errors on non-web
      // ignore: undefined_prefixed_name
      final dynamic html = _getHtmlDocument();
      if (html != null) {
        html.cookie = value;
      }
    } catch (_) {}
  }

  static String? getCookie(String name) {
    if (!kIsWeb) return null;
    try {
      final dynamic html = _getHtmlDocument();
      if (html == null) return null;
      final String cookies = html.cookie ?? '';
      for (final cookie in cookies.split(';')) {
        final parts = cookie.trim().split('=');
        if (parts.length >= 2 && parts[0] == name) {
          return parts.sublist(1).join('=');
        }
      }
    } catch (_) {}
    return null;
  }

  static dynamic _getHtmlDocument() {
    try {
      // Use dart:html document on web
      return _unsafeGetDocument();
    } catch (_) {
      return null;
    }
  }
}

// This function isolates the dart:html dependency
dynamic _unsafeGetDocument() {
  // This will be replaced by the web-specific implementation
  // In web builds, dart:html is available
  // In non-web builds, this throws and is caught
  throw UnsupportedError('Not on web platform');
}

File: frontend/frontend/src/lib/cross-app-auth-bridge.ts
/**
 * Cross-App Auth Bridge — Next.js Side
 *
 * Reads the HMAC-signed session token that Flutter writes to a cookie.
 * This allows Next.js to trust the Flutter session without its own
 * separate localStorage auth system when users navigate to War Room.
 *
 * Cookie name: amttp_xauth
 * Token format: base64url(payload).base64url(hmac-sha256)
 *
 * The HMAC key MUST match Flutter's AUTH_BRIDGE_KEY dart-define.
 */

// Key must match Flutter's --dart-define=AUTH_BRIDGE_KEY value
const BRIDGE_KEY = process.env.AUTH_BRIDGE_KEY || 'amttp-dev-bridge-key-change-in-production-2026';
const COOKIE_NAME = 'amttp_xauth';

// ═══════════════════════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════════════════════

export interface BridgeTokenPayload {
  sub: string;     // userId
  email: string;
  role: string;    // e.g. "R3_INSTITUTION_OPS"
  mode: string;    // "FOCUS" or "WAR_ROOM"
  name: string;    // displayName
  iat: number;     // issued-at (epoch seconds)
  exp: number;     // expiry (epoch seconds)
}

// ═══════════════════════════════════════════════════════════════════════════════
// HMAC UTILITIES (Web Crypto API — works in both Node.js and browser)
// ═══════════════════════════════════════════════════════════════════════════════

function base64UrlEncode(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer);
  let binary = '';
  bytes.forEach(b => binary += String.fromCharCode(b));
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

function base64UrlDecode(str: string): Uint8Array {
  // Restore standard base64
  let base64 = str.replace(/-/g, '+').replace(/_/g, '/');
  while (base64.length % 4) base64 += '=';
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

async function hmacSign(data: string): Promise<string> {
  const enc = new TextEncoder();
  const keyData = enc.encode(BRIDGE_KEY);
  const msgData = enc.encode(data);

  const cryptoKey = await crypto.subtle.importKey(
    'raw', keyData, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']
  );
  const sig = await crypto.subtle.sign('HMAC', cryptoKey, msgData);
  return base64UrlEncode(sig);
}

async function hmacVerify(data: string, signature: string): Promise<boolean> {
  const expected = await hmacSign(data);
  // Constant-time comparison
  if (expected.length !== signature.length) return false;
  let result = 0;
  for (let i = 0; i < expected.length; i++) {
    result |= expected.charCodeAt(i) ^ signature.charCodeAt(i);
  }
  return result === 0;
}

// ═══════════════════════════════════════════════════════════════════════════════
// TOKEN OPERATIONS
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Verify and decode a bridge token.
 * Returns the payload if valid, null if invalid/expired/tampered.
 */
export async function verifyBridgeToken(token: string): Promise<BridgeTokenPayload | null> {
  try {
    const parts = token.split('.');
    if (parts.length !== 2) return null;

    const [payloadB64, sig] = parts;

    // Verify HMAC
    const valid = await hmacVerify(payloadB64, sig);
    if (!valid) {
      console.warn('[CrossAppAuth] HMAC verification failed — possible tampering');
      return null;
    }

    // Decode payload
    const payloadBytes = base64UrlDecode(payloadB64);
    const payloadStr = new TextDecoder().decode(payloadBytes);
    const payload = JSON.parse(payloadStr) as BridgeTokenPayload;

    // Check expiry
    const now = Math.floor(Date.now() / 1000);
    if (now > payload.exp) {
      console.warn('[CrossAppAuth] Token expired');
      return null;
    }

    return payload;
  } catch (e) {
    console.error('[CrossAppAuth] Token verification error:', e);
    return null;
  }
}

/**
 * Create a bridge token (for testing or Next.js-initiated sessions).
 */
export async function createBridgeToken(payload: Omit<BridgeTokenPayload, 'iat' | 'exp'>): Promise<string> {
  const now = Math.floor(Date.now() / 1000);
  const full: BridgeTokenPayload = {
    ...payload,
    iat: now,
    exp: now + 86400, // 24 hours
  };

  const enc = new TextEncoder();
  const payloadB64 = base64UrlEncode(enc.encode(JSON.stringify(full)));
  const sig = await hmacSign(payloadB64);
  return `${payloadB64}.${sig}`;
}

// ═══════════════════════════════════════════════════════════════════════════════
// COOKIE READING (Client-side)
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Read the bridge token from cookie (client-side).
 */
export function readBridgeTokenFromCookie(): string | null {
  if (typeof document === 'undefined') return null;
  const cookies = document.cookie.split(';');
  for (const cookie of cookies) {
    const [name, ...valueParts] = cookie.trim().split('=');
    if (name === COOKIE_NAME) {
      return valueParts.join('=');
    }
  }
  return null;
}

/**
 * Read bridge token from localStorage fallback.
 */
export function readBridgeTokenFromStorage(): string | null {
  if (typeof localStorage === 'undefined') return null;
  return localStorage.getItem('amttp_bridge_token');
}

/**
 * Read bridge token — tries cookie first, then localStorage.
 */
export function readBridgeToken(): string | null {
  return readBridgeTokenFromCookie() || readBridgeTokenFromStorage();
}

/**
 * Get verified session from the bridge token.
 * Returns null if no token, invalid, or expired.
 */
export async function getBridgeSession(): Promise<BridgeTokenPayload | null> {
  const token = readBridgeToken();
  if (!token) return null;
  return verifyBridgeToken(token);
}

/**
 * Check if a bridge session exists and is valid.
 */
export async function hasBridgeSession(): Promise<boolean> {
  const session = await getBridgeSession();
  return session !== null;
}

/**
 * Clear bridge session (logout).
 */
export function clearBridgeSession(): void {
  if (typeof document !== 'undefined') {
    document.cookie = `${COOKIE_NAME}=; path=/; SameSite=Strict; max-age=0`;
  }
  if (typeof localStorage !== 'undefined') {
    localStorage.removeItem('amttp_bridge_token');
  }
}
